File: app/src/test/java/org/metabrainz/mobile/LookupRepositoryTest.java
package org.metabrainz.mobile;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.LiveData;

import com.google.gson.Gson;

import org.jetbrains.annotations.NotNull;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.metabrainz.mobile.data.repository.LookupRepository;
import org.metabrainz.mobile.data.sources.Constants;
import org.metabrainz.mobile.data.sources.api.LookupService;
import org.metabrainz.mobile.data.sources.api.MusicBrainzServiceGenerator;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Artist;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Label;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.MBEntityType;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Recording;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Release;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.ReleaseGroup;

import okhttp3.mockwebserver.Dispatcher;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;

import static org.junit.Assert.fail;
import static org.metabrainz.mobile.AssertionUtils.checkArtistAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkLabelAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkRecordingAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkReleaseAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkReleaseGroupAssertions;
import static org.metabrainz.mobile.EntityTestUtils.getTestArtist;
import static org.metabrainz.mobile.EntityTestUtils.getTestArtistMBID;
import static org.metabrainz.mobile.EntityTestUtils.getTestLabel;
import static org.metabrainz.mobile.EntityTestUtils.getTestLabelMBID;
import static org.metabrainz.mobile.EntityTestUtils.getTestRecording;
import static org.metabrainz.mobile.EntityTestUtils.getTestRecordingMBID;
import static org.metabrainz.mobile.EntityTestUtils.getTestRelease;
import static org.metabrainz.mobile.EntityTestUtils.getTestReleaseGroup;
import static org.metabrainz.mobile.EntityTestUtils.getTestReleaseGroupMBID;
import static org.metabrainz.mobile.EntityTestUtils.getTestReleaseMBID;
import static org.metabrainz.mobile.EntityTestUtils.loadResourceAsString;

public class LookupRepositoryTest {

    @Rule
    public InstantTaskExecutorRule executorRule = new InstantTaskExecutorRule();
    MockWebServer webServer;
    LookupRepository repository;

    @Before
    public void setup() {
        try {
            webServer = new MockWebServer();
            webServer.setDispatcher(new Dispatcher() {
                @NotNull
                @Override
                public MockResponse dispatch(@NotNull RecordedRequest recordedRequest) {
                    String endpoint = recordedRequest.getPath()
                            .substring(1, recordedRequest.getPath().indexOf('/', 1));
                    String file = endpoint + "_lookup.json";
                    return new MockResponse()
                            .setResponseCode(200)
                            .setBody(loadResourceAsString(file));
                }
            });
            webServer.start();
            LookupService service = MusicBrainzServiceGenerator.createTestService(
                    LookupService.class, webServer.url("/"));
            repository = new LookupRepository(service);
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testArtistLookup() {
        Artist testArtist = getTestArtist();
        LiveData<String> testArtistData = repository.fetchData(MBEntityType.ARTIST.name,
                getTestArtistMBID(), Constants.LOOKUP_ARTIST_PARAMS);
        try {
            String response = LiveDataTestUtil.getOrAwaitValue(testArtistData);
            Artist artist = new Gson().fromJson(response, Artist.class);
            checkArtistAssertions(testArtist, artist);
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testReleaseLookup() {
        Release testRelease = getTestRelease();
        LiveData<String> testReleaseData = repository.fetchData(MBEntityType.RELEASE.name,
                getTestReleaseMBID(), Constants.LOOKUP_RELEASE_PARAMS);
        try {
            String response = LiveDataTestUtil.getOrAwaitValue(testReleaseData);
            Release release = new Gson().fromJson(response, Release.class);
            checkReleaseAssertions(testRelease, release);
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testReleaseGroupLookup() {
        ReleaseGroup testReleaseGroup = getTestReleaseGroup();
        LiveData<String> testReleaseGroupData = repository.fetchData(MBEntityType.RELEASE_GROUP.name,
                getTestReleaseGroupMBID(), Constants.LOOKUP_RELEASE_GROUP_PARAMS);
        try {
            String response = LiveDataTestUtil.getOrAwaitValue(testReleaseGroupData);
            ReleaseGroup releaseGroup = new Gson().fromJson(response, ReleaseGroup.class);
            checkReleaseGroupAssertions(testReleaseGroup, releaseGroup);
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testLabelLookup() {
        Label testLabel = getTestLabel();
        LiveData<String> testLabelData = repository.fetchData(MBEntityType.LABEL.name,
                getTestLabelMBID(), Constants.LOOKUP_LABEL_PARAMS);
        try {
            String response = LiveDataTestUtil.getOrAwaitValue(testLabelData);
            Label label = new Gson().fromJson(response, Label.class);
            checkLabelAssertions(testLabel, label);
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testRecordingLookup() {
        Recording testRecording = getTestRecording();
        LiveData<String> testRecordingData = repository.fetchData(MBEntityType.RECORDING.name,
                getTestRecordingMBID(), Constants.LOOKUP_RECORDING_PARAMS);
        try {
            String response = LiveDataTestUtil.getOrAwaitValue(testRecordingData);
            Recording recording = new Gson().fromJson(response, Recording.class);
            checkRecordingAssertions(testRecording, recording);
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @After
    public void teardown() {
        try {
            webServer.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/test/java/org/metabrainz/mobile/LookupRepositoryTest.kt
package org.metabrainz.mobile

import com.google.gson.Gson
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.runBlocking
import okhttp3.mockwebserver.Dispatcher
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okhttp3.mockwebserver.RecordedRequest
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.metabrainz.mobile.AssertionUtils.checkArtistAssertions
import org.metabrainz.mobile.AssertionUtils.checkLabelAssertions
import org.metabrainz.mobile.AssertionUtils.checkRecordingAssertions
import org.metabrainz.mobile.AssertionUtils.checkReleaseAssertions
import org.metabrainz.mobile.AssertionUtils.checkReleaseGroupAssertions
import org.metabrainz.mobile.EntityTestUtils.loadResourceAsString
import org.metabrainz.mobile.EntityTestUtils.testArtist
import org.metabrainz.mobile.EntityTestUtils.testLabel
import org.metabrainz.mobile.EntityTestUtils.testRecording
import org.metabrainz.mobile.EntityTestUtils.testRelease
import org.metabrainz.mobile.EntityTestUtils.testReleaseGroup
import org.metabrainz.mobile.RetrofitUtils.createTestService
import org.metabrainz.mobile.data.repository.LookupRepository
import org.metabrainz.mobile.data.repository.LookupRepositoryImpl
import org.metabrainz.mobile.data.sources.Constants.LOOKUP_ARTIST_PARAMS
import org.metabrainz.mobile.data.sources.Constants.LOOKUP_LABEL_PARAMS
import org.metabrainz.mobile.data.sources.Constants.LOOKUP_RECORDING_PARAMS
import org.metabrainz.mobile.data.sources.Constants.LOOKUP_RELEASE_GROUP_PARAMS
import org.metabrainz.mobile.data.sources.Constants.LOOKUP_RELEASE_PARAMS
import org.metabrainz.mobile.data.sources.api.LookupService
import org.metabrainz.mobile.data.sources.api.entities.mbentity.*
import org.metabrainz.mobile.util.Resource.Status.SUCCESS

class LookupRepositoryTest {

    // FIXME: Replaced runBlockingTest with runBlocking for the time being to pass the tests
    // This is a bug in the kotlinx-coroutines-test library. This should not be necessary once the
    // bug is fixed upstream.
    private lateinit var webServer: MockWebServer
    private lateinit var repository: LookupRepository

    @Before
    fun setup() {
        webServer = MockWebServer()
        webServer.dispatcher = object : Dispatcher() {
            override fun dispatch(request: RecordedRequest): MockResponse {
                val endpoint = request.path?.substring(1, request.path!!.indexOf('/', 1))
                val file = endpoint + "_lookup.json"
                return MockResponse().setResponseCode(200).setBody(loadResourceAsString(file))

                // Head to https://github.com/square/okhttp/tree/master/mockwebserver for documentation
                // Equivalent to
                /*
                    when (request.path) {
                       "/v1/check/version/" -> {
                           return MockResponse().setResponseCode(200).setBody(loadResourceAsString(file))
                       }
                   }
                 */
            }
        }
        webServer.start()
        val service = createTestService(LookupService::class.java, webServer.url("/"))
        repository = LookupRepositoryImpl(service)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testArtistLookup() = runBlocking {
        val expected = testArtist
        val resource = repository.fetchData(MBEntityType.ARTIST.entity, expected.mbid!!, LOOKUP_ARTIST_PARAMS)
        assertEquals(SUCCESS, resource.status)
        val actual = Gson().fromJson(resource.data, Artist::class.java)
        checkArtistAssertions(expected, actual)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testReleaseLookup() = runBlocking {
        val expected = testRelease
        val resource = repository.fetchData(MBEntityType.RELEASE.entity, expected.mbid!!, LOOKUP_RELEASE_PARAMS)
        assertEquals(SUCCESS, resource.status)
        val actual = Gson().fromJson(resource.data, Release::class.java)
        checkReleaseAssertions(expected, actual)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testReleaseGroupLookup() = runBlocking {
        val expected = testReleaseGroup
        val resource = repository.fetchData(MBEntityType.RELEASE_GROUP.entity, expected.mbid!!, LOOKUP_RELEASE_GROUP_PARAMS)
        assertEquals(SUCCESS, resource.status)
        val actual = Gson().fromJson(resource.data, ReleaseGroup::class.java)
        checkReleaseGroupAssertions(expected, actual)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testLabelLookup() = runBlocking {
        val expected = testLabel
        val resource = repository.fetchData(MBEntityType.LABEL.entity, expected.mbid!!, LOOKUP_LABEL_PARAMS)
        assertEquals(SUCCESS, resource.status)
        val label = Gson().fromJson(resource.data, Label::class.java)
        checkLabelAssertions(expected, label)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testRecordingLookup() = runBlocking {
        val expected = testRecording
        val resource = repository.fetchData(MBEntityType.RECORDING.entity, expected.mbid!!, LOOKUP_RECORDING_PARAMS)
        assertEquals(SUCCESS, resource.status)
        val recording = Gson().fromJson(resource.data, Recording::class.java)
        checkRecordingAssertions(expected, recording)
    }

    @After
    fun teardown() {
        webServer.close()
    }
}